File: ext_serial.ts
/**
 * Extensions to the serial module.
 *
 * @author Matthias L. Jugel
 */
namespace serial {
    /**
     * Set serial receive buffer size.
     * @param size the new size of the receive buffer, eg: 20
     */
    //% blockId=serial_buffersize block="serial receive buffer size %size"
    //% shim=serial::setReceiveBufferSize
    export function setReceiveBufferSize(size: number): void {
        return;
    }

    /**
     * Reset serial back to USBTX/USBRX.
     */
    //% blockId=serial_resetserial block="serial reset"
    //% shim=serial::resetSerial
    export function resetSerial(): void {
        return;
    }

    //% shim=serial::read
    export function read(delimiters: string): string {
        return "OK";
    }

    //% shim=serial::busy
    export function busy(): boolean {
        return false;
    }

}

File: ext_serial.cpp
#include "pxt.h"

namespace serial {
    //%
    void setReceiveBufferSize(int size) {
        // make sure we only allocate 255 bytes or the device will freeze
        uBit.serial.setRxBufferSize(size < 255 ? size : 254);
    }

    //%
    bool busy() {
        return uBit.serial.txInUse();
    }

    //%
    void resetSerial() {
        while(uBit.serial.redirect(USBTX, USBRX) == MICROBIT_SERIAL_IN_USE) fiber_sleep(10);
        uBit.serial.baud(MICROBIT_SERIAL_DEFAULT_BAUD_RATE);
    }

    //%
    StringData *read(StringData *delimiters) {
        return uBit.serial.readUntil(ManagedString(delimiters), MicroBitSerialMode::SYNC_SPINWAIT).leakData();
    }
}
